asllib/types.ml

```diff
@@ -651,15 +653,35 @@ and structural_subtype_satisfies env t s =
         every bitfield in S there must be a bitfield in T of the same name, width
         and offset, whose type type-satisfies the bitfield in S.
     *)
-  | T_Bits (w_s, bf_s), T_Bits (w_t, bf_t) -> (
-      (* Interpreting the first two condition as just a condition on
-         domains. *)
-      match (bf_s, bf_t) with
-      | [], _ -> true
-      | _, [] -> false
-      | bfs_s, bfs_t ->
-          bitwidth_equal env w_s w_t && bitfields_included env bfs_s bfs_t)
-  | T_Bits _, _ -> false
+  | T_Bits (_w_s, bf_s), T_Bits (_w_t, bf_t) ->
+      let bitfields_subtype =
+        match (bf_s, bf_t) with
+        | [], _ -> true
+        | _, [] -> false
+        | bfs_s, bfs_t -> bitfields_included env bfs_s bfs_t
+      in
+      let widths_subtype =
+        (*
+        • If either S or T have the structure of a bitvector type with
+          undetermined width then the domain of T must be a subset of the
+          domain of S.
+         *)
+        (* Implicitly, T must have the structure of a bitvector. *)
+        let t_domain = get_structure env t |> Domain.of_type env
+        and s_domain = get_structure env s |> Domain.of_type env in
+        let () =
+          if false then
+            Format.eprintf "Is %a included in %a?@." Domain.pp t_domain
+              Domain.pp s_domain
+        in
+        match
+          ( Domain.get_width_singleton_opt s_domain,
+            Domain.get_width_singleton_opt t_domain )
+        with
+        | Some w_s, Some w_t -> Z.equal w_s w_t
+        | _ -> Domain.is_subset env t_domain s_domain
```

Comment:
Would you like me to add this in? A quick test suggests that this would produce more "interval too large" warnings, but would not affect error behaviour in any tests.

Reply:
No, leave it as is, but thanks for checking. I noted it so we can come back to this later.